### PR TITLE
 spinlock: add support of spin_trylock_irqsave()

### PR DESCRIPTION
**Pull request by anchao**
## Summary

spinlock: add support of spin_trylock_irqsave()
trylock spinlock in critical section:

```
bool spin_trylock_irqsave(FAR volatile spinlock_t *lock, irqstate_t flags);
bool spin_trylock_irqsave_wo_note(FAR volatile spinlock_t *lock, irqstate_t flags);
```

Signed-off-by: chao an <anchao@lixiang.com>

## Impact
N/A
## Testing

ci-check

### CODE REVIEW

**Review by acassis — APPROVED** [review]

File: include/nuttx/spinlock.h [hunk 1]
```diff
@@ -626,6 +626,81 @@ irqstate_t spin_lock_irqsave(FAR volatile spinlock_t *lock)
...
+ *   SP_UNLOCKED - Success, the spinlock was successfully locked
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SPINLOCK
+#  define spin_trylock_irqsave_wo_note(l, f) \
+({ \
```
Comment (xiaoxiang781216) [inline include/nuttx/spinlock.h:654]: but not all compiler support the expression statement.

Comment (anchao) [reply include/nuttx/spinlock.h:654]: Only MSVC does not support this syntax, which is  simulation environment and does not require spinlock support.

**Comment by xiaoxiang781216** [comment]
@anchao please rebase the change

File: include/nuttx/spinlock.h (hunk 1 above)
Comment (xiaoxiang781216) [reply include/nuttx/spinlock.h:654]: if sim on windows support SMP, this code will break windows porting.

Comment (anchao) [reply include/nuttx/spinlock.h:654]: There is already a similar implement in sim, so are xiaomi still using MSVC in windows?
https://github.com/apache/nuttx/blob/master/arch/sim/src/sim/sim_internal.h#L127-L135

Comment (anchao) [reply include/nuttx/spinlock.h:654]: There is already a similar implement in sim, so are xiaomi still using MSVC in windows?
https://github.com/apache/nuttx/blob/master/arch/sim/src/sim/sim_internal.h#L127-L135

Comment (xiaoxiang781216) [reply include/nuttx/spinlock.h:654]: I don' t see there is any issue to block use inline function.

Comment (anchao) [reply include/nuttx/spinlock.h:654]: If similar code in xiaomi internal repository can be compiled normally, it means that there is no issue with the `expression statement` on MSVC. could we merge this PR ?

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**